src/algebra/cubic_solution.lean

```diff
@@ -0,0 +1,570 @@
+/-
+Copyright (c) 2021 Nicholas Dyson. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Nicholas Dyson.
+-/
+
+import algebra.invertible
+import tactic.linarith
+import tactic.field_simp
+import order.filter.at_top_bot
+
+/-!
+# Cubic solutions
+
+In this file we define and prove the three solutions to a general cubic in a general
+field that is not of characteristic two or three, and uniqueness of these solutions.
+
+## Main results
+
+`cubic_solution` : The solution to a cubic
+`cubic_solution_unique` : The three solutions generated by using the three cube roots in the
+                          formula are the only solutions.
+-/
+
+
+section fields
+variables {F : Type*}
+
+/--
+((-1 + sqrt(-3))/2)^3 = 1
+-/
+def cube_root_unity_a [field F] (root_three_i : F) : F :=
+  (-1 + root_three_i)/2
+
+/--
+((-1 - sqrt(-3))/2)^3 = 1
+-/
+def cube_root_unity_b [field F] (root_three_i : F) : F :=
+  (-1 - root_three_i)/2
+
+lemma multiply_out_cubed [field F] (x y : F) : (x + y) * (x + y) * (x + y) =
+        (x * x * x) + (3 * x * x * y) + (3 * x * y * y) + (y * y * y) :=
+begin
+  ring,
+end
+
+/--
+If F isn't of characteristic 2, 2^3 ≠ 0 also.
+--/
+instance four_ne_zero [field F] [invertible (2 : F)] : invertible (4 : F) :=
+begin
+    have cubed : (2 : F) ^ 2 = (4 : F),
+      ring,
+    rw <- cubed, exact invertible_pow 2 2,
+end
+
+/--
+If F isn't of characteristic 2, 2^3 ≠ 0 also.
+--/
+instance eight_ne_zero [field F] [invertible (2 : F)] : invertible (8 : F) :=
+begin
+    have cubed : (2 : F) ^ 3 = (8 : F),
+      ring,
+    rw <- cubed, exact invertible_pow 2 3,
+end
```

Comment:
Untested, but I think this works as:
```suggestion
begin
  convert invertible_pow 2 3,
  ring,
end
```
Note that mathlib style is to indent two spaces not four.

Reply:
This fails with an error I've never seen before (`invalid numeral, failed to force numeral to be a nat`).  If I specify all the types, as in
`convert invertible_pow F (2 : F) (3 : F)`
then I get the following error:
`failed to synthesize type class instance for`
`F : Type u_1,`
`_inst_1 : field F,`
`_inst_2 : invertible 2`
`⊢ monoid (Type u_1)`
which I really don't understand - multiplication in a field has to be a monoid.  Advice would be appreciated.

Reply:
This fails with an error I've never seen before (`invalid numeral, failed to force numeral to be a nat`). If I specify all the types, as in
`convert invertible_pow F (2 : F) (3 : F)`
then I get the following error:
`failed to synthesize type class instance for`
`F : Type u_1,`
`_inst_1 : field F,`
`_inst_2 : invertible 2`
`⊢ monoid (Type u_1)`
which I really don't understand - multiplication in a field has to be a monoid. Advice would be appreciated.

Reply:
You passed `F : Type u_1` into an argument that was expecting `2: F`. Your error message says "I can't find a monoid structure on `Type u_1`", not "I can't find a monoid structure on `F`".